src/commands/run.rs

```diff
@@ -220,79 +236,40 @@ impl RunCommand {
         result
     }
 
-    fn instantiate_module(
-        store: &Store,
-        module_registry: &ModuleRegistry,
-        path: &Path,
-    ) -> Result<Instance> {
-        // Read the wasm module binary either as `*.wat` or a raw binary
-        let data = wat::parse_file(path)?;
-
-        let module = Module::new(store, &data)?;
-
-        // Resolve import using module_registry.
-        let imports = module
-            .imports()
-            .map(|i| {
-                let export = match i.module() {
-                    "wasi_snapshot_preview1" => {
-                        module_registry.wasi_snapshot_preview1.get_export(i.name())
-                    }
-                    "wasi_unstable" => module_registry.wasi_unstable.get_export(i.name()),
-                    other => bail!("import module `{}` was not found", other),
-                };
-                match export {
-                    Some(export) => Ok(export.clone().into()),
-                    None => bail!(
-                        "import `{}` was not found in module `{}`",
-                        i.name(),
-                        i.module()
-                    ),
-                }
-            })
-            .collect::<Result<Vec<_>, _>>()?;
-
-        let instance = Instance::new(&module, &imports)
-            .context(format!("failed to instantiate {:?}", path))?;
-
-        Ok(instance)
-    }
-
-    fn handle_module(&self, store: &Store, module_registry: &ModuleRegistry) -> Result<()> {
+    fn load_main_module(&self, linker: &mut Linker) -> Result<()> {
         if let Some(timeout) = self.wasm_timeout {
-            let handle = store.interrupt_handle()?;
+            let handle = linker.store().interrupt_handle()?;
             thread::spawn(move || {
                 thread::sleep(timeout);
                 handle.interrupt();
             });
         }
-        let instance = Self::instantiate_module(store, module_registry, &self.module)?;
+
+        // Read the wasm module binary either as `*.wat` or a raw binary.
+        // Use "" as a default module name.
+        let module = Module::from_file(linker.store(), &self.module)?;
+        linker
+            .module("", &module)
```

Comment:
While probably not a huge issue it's a bit odd that we have to assign *a* name to the instance here, since ideally we'd pull out an instance and "do the wasi thing" on it.

Reply:
I used a name here because I think src/commands/run.rs shouldn't be pulling out the instance and doing the initialization manually. The `Linker` handles that for it, which is good because then the functionality is available to API users as well as command-line users.